### ISSUE TITLE
running slow in search of jsonb data type, about 8million rows

### ISSUE THREAD

**Issue by zyp-rgb** [pgroonga/pgroonga#145]
Thanks in advance! I am using pgroonga to search in a jsonb column with around 8 millions rows. It took about 3 seconds to return. Is that normal? Index creating and sql are the followings.

```sql
CREATE INDEX "data-test_extra_idx" ON "data-test"(extra pgroonga_jsonb_ops_v2);

select * FROM "data-test"
WHERE extra &` 'paths @ "district" && string == "shanghai"' 
LIMIT 1000 
```

**Comment by komainu8** [comment]
Thank you for your report.

Could you teach us the result of the following SQL, the table schema, and version of PGroonga and PostgreSQL?

```sql
EAPLAIN ANALYZE VERBOSE select * FROM "data-test"
WHERE extra &` 'paths @ "district" && string == "shanghai"'
LIMIT 1000
```

**Comment by zyp-rgb** [comment]
Limit  (cost=0.00..35555.67 rows=1000 width=301) (actual time=2949.969..2950.539 rows=1000 loops=1)
  ->  Index Scan using "data-eaeefe70-354a-4fdf-b969-92ba6427d61e_extra_idx" on "data-eaeefe70-354a-4fdf-b969-92ba6427d61e"  (cost=0.00..300729.84 rows=8458 width=301) (actual time=2949.966..2950.435 rows=1000 loops=1)
        Index Cond: (extra &` 'paths @ "model_name" && string == "上海网格总量预测模型"'::text)
Planning Time: 0.803 ms
Execution Time: 2963.236 ms

**Comment by zyp-rgb** [comment]
Postgresql 12, and pgroonga 2.2.6

> Thank you for your report.
> 
> Could you teach us the result of the following SQL, the table schema, and version of PGroonga and PostgreSQL?
> 
> ```sql
> EXPLAIN ANALYZE VERBOSE select * FROM "data-test"
> WHERE extra &` 'paths @ "district" && string == "shanghai"'
> LIMIT 1000
> ```

Limit (cost=0.00..35555.67 rows=1000 width=301) (actual time=2949.969..2950.539 rows=1000 loops=1)
-> Index Scan using "data-eaeefe70-354a-4fdf-b969-92ba6427d61e_extra_idx" on "data-eaeefe70-354a-4fdf-b969-92ba6427d61e" (cost=0.00..300729.84 rows=8458 width=301) (actual time=2949.966..2950.435 rows=1000 loops=1)
Index Cond: (extra &` 'paths @ "model_name" && string == "上海网格总量预测模型"'::text)
Planning Time: 0.803 ms
Execution Time: 2963.236 ms

used a differenct cond but should be the same
PostgreSQL 12
pgroonga 2.2.6

**Comment by komainu8** [comment]
Probably, the cause of this problem is the expression of `string == "上海网格总量预测模型"` is executed by a sequential search.

Could you check to see if the following query improves to response time or not?

```sql
SELECT * FROM "data-test"
WHERE extra &` ('paths @ "district" && string @~ ' || groonga_escape('\A' || pgroonga_normalize('shanghai') || '\z'))
LIMIT 1000;
```

If a search keyword includes `.`, we also escape `.` as below.

```sql
SELECT * FROM "data-test"
WHERE extra &` ('paths @ "district" && string @~ ' || groonga_escape('\A' || pgroonga_normalize('www\.example\.com') || '\z'))
LIMIT 1000;
```


---

If this query is slow yet, could you provide Groonga's query log?

We can get Groonga's query log as a below method.


1. We execute the following command on PostgreSQL.
 
```sql
SET pgroonga.query_log_path = 'pgroonga.query.log';
```

2. We execute the following two slow queries.

```sql
SELECT * FROM "data-test"
WHERE extra &` ('paths @ "district" && string @~ ' || groonga_escape('\A' || pgroonga_normalize('shanghai') || '\z'))
LIMIT 1000;
```

```sql
SELECT * FROM "data-test"
WHERE extra &` 'paths @ "district" && string == "shanghai"'
LIMIT 1000
```

3. Could you teach us the contents of `$PGDATA/pgroonga.query.log`?

**Comment by zyp-rgb** [comment]
> ```sql
> SELECT * FROM "data-test"
> WHERE extra &` ('paths @ "district" && string @~ ' || groonga_escape('\A' || pgroonga_normalize('shanghai') || '\z'))
> LIMIT 1000;
> ```

Thanks for the guide. The escape and normalize didn't work. 3 seconds still

I executed two sql. Logs are the followings. I am a newbee but it seems to filter from a large amount of 1215632 while I limit 1000.

SELECT * FROM "data-test"
WHERE extra &` ('paths @ "model_name" && string @~ ' || pgroonga_escape('\A' || pgroonga_normalize('上海网格总量预测模型') || '\z'))
LIMIT 1000;

select * FROM "data-test"
WHERE extra &` 'paths @ "model_name" && string == "上海网格总量预测模型"'
LIMIT 1000

2020-09-24 03:08:45.996948|0x7fcac434d4c0| query log will be closed: <pgroonga.query.log>
2020-09-24 03:08:45.997063|0x7fcac434d4c0| query log is opened: <pgroonga.query.log>
2020-09-24 03:08:52.816365|0x7fcac434d4c0|:000174393775902 filter(16): JSONValues95545_0.paths match "model_name"
2020-09-24 03:08:52.816465|0x7fcac434d4c0|:000174393863153 filter(1): JSONValues95545_0.string "\\A上海网格总量预测模型\\z" regexp
2020-09-24 03:08:55.693690|0x7fcac434d4c0|:000177271098014 filter(1215632): sub_filter(Sources95545.extra, "paths @ \"model_name\" && string @~ \"\\\\A上海网格总量预测模型\\\\z\"")
2020-09-24 03:09:04.626779|0x7fcac434d4c0|:000186204188991 filter(16): JSONValues95545_0.paths match "model_name"
2020-09-24 03:09:04.633952|0x7fcac434d4c0|:000186211386931 filter(1): JSONValues95545_0.string equal "上海网格总量预测模型"
2020-09-24 03:09:07.484678|0x7fcac434d4c0|:000189062086365 filter(1215632): sub_filter(Sources95545.extra, "paths @ \"model_name\" && string == \"上海网格总量预测模型\"")

**Comment by komainu8** [comment]
Ah...Sorry.

We need to set the tokenizer for regex to index.
Could you try again the below queries? 

```sql
DROP INDEX data-test_extra_idx;
CREATE INDEX "data-test_extra_idx" ON "data-test"(extra pgroonga_jsonb_ops_v2) WITH (tokenizer='TokenRegexp');

SELECT * FROM "data-test"
WHERE extra &` ('paths @ "district" && string @~ ' || groonga_escape('\A' || pgroonga_normalize('shanghai') || '\z'))
LIMIT 1000;
```

**Comment by zyp-rgb** [comment]
> Ah...Sorry.
> 
> We need to set the tokenizer for regex to index.
> Could you try again the below queries?
> 
> ```sql
> DROP INDEX data-test_extra_idx;
> CREATE INDEX "data-test_extra_idx" ON "data-test"(extra pgroonga_jsonb_ops_v2) WITH (tokenizer='TokenRegexp');
> 
> SELECT * FROM "data-test"
> WHERE extra &` ('paths @ "district" && string @~ ' || groonga_escape('\A' || pgroonga_normalize('shanghai') || '\z'))
> LIMIT 1000;
> ```

yeah I tried, still the same. 
I use this to recreate the index, just in case of any difference.
And I check the log, the same as above.

CREATE INDEX "data-test_extra_idx" ON "data-test" USING pgroonga (extra pgroonga_jsonb_ops_v2) WITH (tokenizer='TokenRegexp');

**Comment by kou** [comment]
Thanks.

It seems that this `sub_filter(Sources95545.extra, "paths @ "model_name" && string == "上海网格总量预测模型"")` internal operation is bottleneck.

We need to improve performance of the operation in Groonga. We'll try it in a few months. Please wait for a while.

**Comment by zyp-rgb** [comment]
> Thanks.
> 
> It seems that this `sub_filter(Sources95545.extra, "paths @ "model_name" && string == "上海网格总量预测模型"")` internal operation is bottleneck.
> 
> We need to improve performance of the operation in Groonga. We'll try it in a few months. Please wait for a while.

Thanks. Looking forward to it!

**Comment by kou** [comment]
This will be improved by the next Groonga (10.0.8).

**Comment by komainu8** [comment]
@zyp-rgb 
We have released Groonga 10.0.8.
Could you update your Groonga version and try again?

**Comment by komainu8** [comment]
No response.
